corehq/apps/userreports/reports/filters/choice_providers.py

```diff
@@ -86,6 +86,9 @@ def get_sorted_choices_for_values(self, values, user):
 
     def get_choices_for_values(self, values, user):
         choices = set(self.get_choices_for_known_values(values, user))
+        if self.location_safe and not user.has_permission(self.domain, 'access_all_locations'):
```

Comment:
After reading @esoergel 's suggestion [here](https://github.com/dimagi/commcare-hq/pull/32964#discussion_r1324882041) I decided to test this out and it works just fine for solving the problem in this PR. What I also like about this particular solution is the fact that it addresses the problem on a deeper level such that all choice providers extending this class would now benefit from this solution.

Reply:
Thanks @Charl1996.
Even thought this solves the issue at hand, the code here is still very confusing to me.

I am quite confused by this function to start with. I am not clear why it would add unknown values anyway.
I tracked this down to [this](https://github.com/dimagi/commcare-hq/pull/9365/commits/51a550d3645e6f8e553aa34ff580de833569cd8a) commit where it was added just to `UserChoiceProvider` first but then moved one level up to the AbstractClass in [this](https://github.com/dimagi/commcare-hq/pull/9365/commits/5b4bec488b0113f7ac413663dbe1349c2d63e6db) commit later. I can't understand the reason why this would be done at all and that is where the security issue happens.
@dannyroberts I know this is way back, but can you help understand why we would include unknown values as choices? My guess is to provide choice for users who are deleted/archived and that extends to other choice providers also like locations? Hence moved it to Abstract Class?
This class could really use some documentation if we can add it now.

I believe Danny's work was done only for users with full access but it was later when `LocationProvider` was made location safe [here](https://github.com/dimagi/commcare-hq/pull/13476/files#diff-28f12eee0e1ae76fb68c35af4f1233e0b835a2a0c3237ca19f781efad220e404R162) it handled the locations query for initial choices but didn't account for missing values getting added back by the parent class, which added inaccessible locations as well.

I am not against the change here but just need to understand the code around it better.

Reply:
> all choice providers extending this class would now benefit from this solution

Currently there is only one class that has location_safe True and that is the `LocationProvider`, so currently other Choice Providers are unaffected which is good since that reduces the impact of this change.